### PR TITLE
Add support for classpath resources

### PR DESCRIPTION
**Pull request by jjudd**
Some third party libraries expect resources to be found on the classpath at the root of the tree. For example, as exactly "myresource.json". This commit adds support for those resources.

classpath_resources are only allowed on binaries, to avoid namespace conflicts. Just like the Bazel Java rules.

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by jjudd** [comment]
When building Play framework projects we ran into an issue where Play expected things like `application.conf` to be found on the classpath as exactly `application.conf`. This is what we did to get that working.

**Comment by ittaiz** [comment]
Thanks!
I'll take a look soon (today/tomorrow hopefully)

**Comment by ittaiz** [comment]
And sorry for the delay

**Comment by ittaiz** [comment]
test this please (ci bot)

**Comment by ittaiz** [comment]
actually, what do you think about also calling your new binary in https://github.com/bazelbuild/rules_scala/blob/master/test/BUILD#L309? it makes sure scala_binary works in test environment

File: src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java [hunk 1]
```diff
@@ -72,10 +72,17 @@ public void processRequest(List<String> args) throws Exception {
...
+
       /**
        * Extract and copy resources from resource jars
        */
       copyResourceJars(ops.resourceJars, tmpPath);
+
+      /**
+       * Copy classpath resources to root of jar
+       */
+      copyClasspathResources(ops.classpathResourceFiles, tmpPath);
```
[the remark this answers is not in the document]
Comment (jjudd) [reply src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java]: I like it. Will change.

File: src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java [hunk 2]
```diff
@@ -316,6 +324,26 @@ private static void copyResources(
...
+    Path dest
+  ) throws IOException {
+    for(String s : classpathResourceFiles) {
+      Path source = Paths.get(s);
+      Path target = dest.resolve(source.getFileName());
+
+      if(Files.exists(target)) {
+        throw new RuntimeException(
+          "Classpath resource file " + source.getFileName()
+          + "has a namespace conflict with another file: " + target.getFileName()
```
[the remark this answers is not in the document]
Comment (jjudd) [reply src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java]: Nice catch. Will change.

File: test_expect_failure/classpath_resources/ObjectWithDuplicateClasspathResources.scala [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package scala.test.classpathresources
+
+object ObjectWithDuplicateClasspathResources extends App {
+  System.exit(0)
```
[the remark this answers is not in the document]
Comment (jjudd) [reply test_expect_failure/classpath_resources/ObjectWithDuplicateClasspathResources.scala]: I suppose not. Will change.

**Comment by jjudd** [comment]
I'm happy to move the binaries. I'll take a look at that in the next couple days. 

I haven't thought about concatenating resources. I'll think about and get back to you.

**Comment by ittaiz** [comment]
@jjudd I wasn't clear I'm sorry. I meant to *additionally* call your new binary in that line. The code as-is is good, the additional call is just another check

**Comment by ittaiz** [comment]
@jjudd friendly ping

**Comment by jjudd** [comment]
Alright. I finally had a chance to get back to this. I addressed the review comments and made classpath_resource duplicates a warning rather than an error, which is what Bazel's Java rules does: https://github.com/bazelbuild/bazel/blob/master/src/java_tools/singlejar/java/com/google/devtools/build/singlejar/SingleJar.java#L202

**Comment by ittaiz** [comment]
Thanks!

**Pull request merged**